src/passes/TrapMode.cpp

```diff
@@ -0,0 +1,319 @@
+  if (generated.hasFunction(name)) {
+    return;
+  }
+  generated.addFunction(generateBinaryFunc(wasm, curr));
+}
+
+void ensureUnaryFunc(Unary *curr, Module& wasm,
+                     GeneratedTrappingFunctions &generated) {
+  Name name = getUnaryFuncName(curr);
+  if (generated.hasFunction(name)) {
+    return;
+  }
+  generated.addFunction(generateUnaryFunc(wasm, curr));
+}
+
+void ensureF64ToI64JSImport(GeneratedTrappingFunctions &generated) {
+  if (generated.hasImport(F64_TO_INT)) {
+    return;
+  }
+
+  Module& wasm = generated.getModule();
+  auto import = new Import; // f64-to-int = asm2wasm.f64-to-int;
+  import->name = F64_TO_INT;
+  import->module = ASM2WASM;
+  import->base = F64_TO_INT;
+  import->functionType = ensureFunctionType("id", &wasm)->name;
+  import->kind = ExternalKind::Function;
+  generated.addImport(import);
+}
+
+Expression* makeTrappingBinary(Binary* curr, GeneratedTrappingFunctions &generated) {
+  Name name = getBinaryFuncName(curr);
+  if (!name.is() || generated.getMode() == TrapMode::Allow) {
+    return curr;
+  }
+
+  // the wasm operation might trap if done over 0, so generate a safe call
+  WasmType type = curr->type;
+  Module& wasm = generated.getModule();
+  Builder builder(wasm);
+  ensureBinaryFunc(curr, wasm, generated);
+  return builder.makeCall(name, {curr->left, curr->right}, type);
+}
+
+Expression* makeTrappingUnary(Unary* curr, GeneratedTrappingFunctions &generated) {
+  Name name = getUnaryFuncName(curr);
+  TrapMode mode = generated.getMode();
+  if (!name.is() || mode == TrapMode::Allow) {
+    return curr;
+  }
+
+  Module& wasm = generated.getModule();
+  Builder builder(wasm);
+  // WebAssembly traps on float-to-int overflows, but asm.js wouldn't, so we must do something
+  // We can handle this in one of two ways: clamping, which is fast, or JS, which
+  // is precisely like JS but in order to do that we do a slow ffi
+  // If i64, there is no "JS" way to handle this, as no i64s in JS, so always clamp if we don't allow traps
+  if (curr->type != i64 && mode == TrapMode::JS) {
+    // WebAssembly traps on float-to-int overflows, but asm.js wouldn't, so we must emulate that
+    ensureF64ToI64JSImport(generated);
+    Expression* f64Value = ensureDouble(curr->value, wasm.allocator);
+    return builder.makeCallImport(F64_TO_INT, {f64Value}, i32);
+  }
+
+  ensureUnaryFunc(curr, wasm, generated);
+  return builder.makeCall(name, {curr->value}, curr->type);
+}
+
+struct TrapModePass : public WalkerPass<PostWalker<TrapModePass>> {
+public:
+
+  // Needs to be non-parallel so that visitModule gets called after visiting
+  // each node in the module, so we can add the functions that we created.
+  bool isFunctionParallel() override { return false; }
+
+  TrapModePass(TrapMode mode) : mode(mode) {
+    assert(mode != TrapMode::Allow);
+  }
+
+  Pass* create() override { return new TrapModePass(mode); }
+
+  void visitUnary(Unary* curr) {
+    ensureGenerated();
```

Comment:
instead of `ensureGenerated` in each expression visitor, can implement `doWalkModule`, and in it call `ensureGenerated()` before calling the ancestor `doWalkModule`.